sdks/java/io/kinesis/src/test/java/org/apache/beam/sdk/io/kinesis/ShardReadersPoolTest.java

```diff
@@ -324,10 +325,15 @@ public void shouldCallRateLimitPolicy()
       }
     }
 
+    ArgumentCaptor<List<KinesisRecord>> recordsCaptor = ArgumentCaptor.forClass(List.class);
+    verify(customRateLimitPolicy, atLeastOnce()).onSuccess(recordsCaptor.capture());
+    List<List<KinesisRecord>> capturedRecords = recordsCaptor.getAllValues();
+    assertThat(capturedRecords).contains(
+        ImmutableList.of(a, b),
+        singletonList(c),
+        singletonList(d),
+        Collections.emptyList()
+    );
     verify(customRateLimitPolicy).onThrottle(same(e));
-    verify(customRateLimitPolicy).onSuccess(eq(ImmutableList.of(a, b)));
-    verify(customRateLimitPolicy).onSuccess(eq(singletonList(c)));
-    verify(customRateLimitPolicy).onSuccess(eq(singletonList(d)));
```

Comment:
Would you share your thought on in what situation (the order of concurrent operations) customRateLimitPolicy did not take `[d]` even when it takes `[a, b]` and `[c]`?

https://builds.apache.org/job/beam_PreCommit_Java_Commit/10277/testReport/junit/org.apache.beam.sdk.io.kinesis/ShardReadersPoolTest/shouldCallRateLimitPolicy/

> Argument(s) are different! Wanted:
> customRateLimitPolicy.onSuccess(
>    [d]
> );
> -> at org.apache.beam.sdk.io.kinesis.ShardReadersPoolTest.shouldCallRateLimitPolicy(ShardReadersPoolTest.java:330)

Reply:
@suztomo My thought is that Mockito is expecting to see these parameters in this order, which may not always be the case due to the 2 ShardReadersPool worker threads. My suggested fix is to rewrite the Mockito verifications so that they are definitely not dependent on order.

Reply:
Thank you for response. I still need your help here. Mockito is not supposed to check the order of the method calls (unless [configured to use inOrde](https://stackoverflow.com/questions/21901368/mockito-verify-order-sequence-of-method-calls)r). I checked its behavior with the following code:
<img width="1019" alt="Screen Shot 2020-03-11 at 12 03 34 AM" src="https://user-images.githubusercontent.com/28604/76381431-da928380-632b-11ea-97d1-0fc0bac12083.png">

Do you know something special in this test case that makes mockito's `verify` order sensitive?

Reply:
Thanks, that's how I thought Mockito.verify() worked too but this made me doubt it. I didn't do any detailed testing, though, so thanks for taking the time to validate. I'm honestly at a loss for what could be causing intermittent failures then. All I can say is this version of the test is semantically equivalent and might not be subject to the same failure mode. I have not been able to reproduce the failure myself.

Reply:
How about writing one-off code that invokes the test, say, 100 times and count test failures?  We can verify whether your version fixes the flakiness.

I don’t expect that one-off code to be checked in, but attaching the observation and the code snippet to PR is helpful.

Reply:
I did run this test thousands of times locally (with `@RepeatedTest(n)` and some hacks to run it on Junit5) and I've managed to reproduce it only twice with the similar but a bit different error as it was on Jenkins.  

Jenkins error: https://gist.github.com/aromanenko-dev/f9714c3dd5a47a9439b9245f3ef7b04b
My local error (old test version): https://gist.github.com/aromanenko-dev/9a8acf265b96e010ad0ad1d09bb3ab92

The common thing there is that it starts to emit empty records list after it returned a list with `c` value.

Also, I did the same thing for new version of test and it's also reproducible (2 fails over 1000 runs):
My local error (new test version) 1st fail: https://gist.github.com/aromanenko-dev/faa4229a486117d7504e486282503009
My local error (new test version) 2nd fail: https://gist.github.com/aromanenko-dev/751cd007bb89a4f6e67f1ce9f9ef3d25

As you can see the actually the errors with old and new versions tests are similar. So, it seems that it's not related to `verify` using and more related to how every mock `ShardRecordsIterator` returns the records. For some reasons it starts to return just empty list.

Reply:
@aromanenko-dev Thank you for sharing your result.

Reply:
Yes, thanks Alexey. I was just working on something similar. Maybe the test is just overspecified then. I propose we change the validations to just:
```
verify(customRateLimitPolicy).onThrottle(any(KinesisClientThrottledException.class));
verify(customRateLimitPolicy, atLeastOnce()).onSuccess(any(List.class));
```
If you both agree I'll make that change.

Reply:
I disagree (I’m not a committer). We just revealed a bug, which is very hard to reproduce.

@aromanenko-dev 

> @RepeatedTest(10000) and some hacks to run it on Junit5

Would you share this?

Reply:
> Would you share this?
 +1, please

@suztomo What bug are you suggesting this exposes?

Reply:
bug: sometimes customRateLimitPolicy does not receive items, "[d]" in the test case.

Reply:
@suztomo You have a good point. The intention of this test is to validate that every time readNextBatch() returns a list of records onSuccess() gets called and every time it throws `KinesisClientThrottledException` onThrottle() gets called. I can see a couple of code paths where readNextBatch() could return records without onSuccess() getting called, namely if any exceptions are thrown along in here: https://github.com/apache/beam/blob/d62521f69ead4b58924043f041978d49e9beeb62/sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java#L146-L149

I haven't seen evidence that this is what's happening, but let's try my latest changes in ShardReadersPool and see if that fixes it. With this change it should not be possible for readNextBatch() to return without onSuccess() being called.

Reply:
@aromanenko-dev If you could share with us how you got @RepeatedTest working I'd be happy to retest.

Reply:
Sure, this patch is enough to run only this test `shouldCallRateLimitPolicy()` with Junit 5:
https://gist.github.com/aromanenko-dev/02c4b206bd7c64bec3ee7a4e3588a5b6
I didn't test other tests, so plz don't blame if something doesn't work =) Sorry for delay, I'm in a different time zone.

Reply:
@aromanenko-dev Thank you very much. I can now reproduce the problem.

![image](https://user-images.githubusercontent.com/28604/76535379-23902800-6451-11ea-8803-e0aedd85c15b.png)

Everyone is in different time zones.

Reply:
What I can guess for now is that `firstIterator` and `secondIterator` are executed in different threads (since `readLoop()` is called for every topic shard in the separate thread) and for some reasons **sometimes** one thread takes more priority over the second while running and the `secondIterator` doesn't have a chance to return "d" record since `shardReadersPool` got closed.
PS: Interesting that on Jenkins it's much easier reproducible than locally.

Reply:
I did some testing on my side and I think that the problem is with using the same mock of `RateLimitPolicy` for both `ShardRecordsIterator`s. For example, when failed assertion says that "d" record is absent in `capturedRecords`, actually I see (using debug output) that it was read with `shardRecordsIterator.readNextBatch()` in `readLoop()`.
Since, in case of using real `RateLimitPolicy` they will be created per `readLoop()` thread (right?), then we need to test it in the same way with mock policies.

Reply:
OK, I see what's happening here. It's basically a race condition. This loop will exit as soon as nextRecord() returns the last record:
https://github.com/apache/beam/blob/b0e7afbc3baf4ac17f7159cb8aa1fabe326be3e2/sdks/java/io/kinesis/src/test/java/org/apache/beam/sdk/io/kinesis/ShardReadersPoolTest.java#L320-L325

I believe that sometimes the test is hitting the verify() call for that last record in the split second between shardReadersPool putting the record into the blocking queue and calling onSuccess(). The fix is to use Mockito's timeout() to ensure that we wait long enough for that last onSuccess() call. It took me about 30,000 iterations to reproduce with the original code but I have run the new test code over 100,000 times without a failure.

I do still think it's a good idea to make these changes to ShardReadersPool to ensure that onSuccess() always gets called (as long as the thread wasn't interrupted).

Reply:
> Since, in case of using the real RateLimitPolicy in normal code, the instances of them will be created per every readLoop() thread (right?). So, we need to test it in the same way with mock policies.

It's possible for a custom RateLimitPolicyFactory to return a singleton instance of RateLimitPolicy, and I think that should be fine (as long as it's thread-safe). So maybe we shouldn't assume that they will always be different instances.